unblob/handlers/archive/zip.py

```diff
@@ -98,78 +98,76 @@
 MAXIMUM_VERSION = 0xFF
 
 
-def _find_end_of_zip(file: io.BufferedIOBase, start_offset: int) -> int:
-    """Find the end of the zip file
-    by looking for the end of central directory header bytes, verifying, then
-    returning the end of the end of central directory header structure.
-    """
-    file.seek(start_offset)
-    content = file.read()
-    end_marker = content.find(b"\x50\x4b\x05\x06")
-    if end_marker == -1:
-        logging.debug(
-            f"ZIP (0x{start_offset:x}): No End of Central Directory headers in the rest of the stream."
-        )
-        return 0
-
-    file.seek(start_offset + end_marker)
-    header = cparser.end_of_central_directory(file)
-
-    try:
-        header.zip_file_comment.decode("utf-8")
-    except UnicodeDecodeError:
-        return _find_end_of_zip(file, start_offset + end_marker + 22)
-
-    return start_offset + end_marker + len(header)
-
-
-def _guess_zip_size(file: io.BufferedIOBase, start_offset: int):
-    # If we just pass a full firmware blob to zipfile.ZipFile, somehow,
-    # the way that it is parsed means that only the final zipfile in the
-    # blob is recognised, if at all. Sometimes, if the firmware is just
-    # a big blob of lots of other things, then ZipFile will just throw an
-    # error. Basically, ZipFile is really bad at dealing with anything
-    # which isn't actually a ZIP.
-
-    # For this reason, we need to try to guess the length of the ZIP file
-    # chunk within our firmware image, and then carve that chunk out.
-    # Then, we make this is a BytesIO stream, so we can just pass this
-    # stream to ZipFile.
-
-    file_names = set()
-    encrypted_files = set()
-    zip_end = _find_end_of_zip(file, start_offset)
+def calculate_zipfile_end(file: io.BufferedIOBase, start_offset: int) -> int:
+    # If we just pass a firmware blob with multiple ZIP files in it to zipfile.ZipFile, it seems
+    # that it will basically scan for the final EOCD record header, and assume that that's where
+    # the file ends.
+    # E.g. in the case our firmware image looks like this:
+    # | ZIPFILE | SOMETHING ELSE | ZIPFILE |
+    # zipfile.ZipFile() will assume:
+    # |    THIS IS ALL THE SAME ZIPFILE    |
+    # For obvious reasons, this is not helpful in our case. We need to try to guess the length of
+    # the ZIP file chunk within our firmware image, independently, and then carve that chunk out.
 
     file.seek(start_offset)
-    content = io.BytesIO(file.read(zip_end - start_offset))
 
-    with ZipFile(content) as z:
-        logger.info("Found ZIP filenames", filenames=[x.filename for x in z.infolist()])
-        for g in z.infolist():
-            if g.flag_bits & 0b0001:
-                encrypted_files.add(g.filename)
-            file_names.add(g.filename)
+    # In our case, we want to find the first instance of the EOCD record header, not the last!
+    zip_end = find_first(file, b"\x50\x4b\x05\x06") + start_offset
```

Comment:
Please make things explicit by using a variable name explaining what we are actually looking for.

```suggestion
    eocd_marker = b"\x50\x4b\x05\x06"
    zip_end = find_first(file, eocd_marker) + start_offset
```